src/main/scala/leon/purescala/Constructors.scala

```diff
@@ -144,7 +145,7 @@ object Constructors {
 
     resType match {
       case Some(tpe) =>
-        casesFiltered.filter(c => isSubtypeOf(c.rhs.getType, tpe) || isSubtypeOf(tpe, c.rhs.getType))
+        casesFiltered.filter(c => isSubtypeOf(c.rhs.getType, tpe) || isSubtypeOf(tpe, c.rhs.getType) || c.optGuard.nonEmpty)
```

Comment:
Why this extra condition? We should not keep cases that do not type-check only because there is a guard? I'm guessing this was introduced as a pseudo-fix to the issue with swapping guard and rhs?